src/mongocrypt.c

```diff
@@ -574,30 +530,403 @@ mongocrypt_init (mongocrypt_t *crypt)
+ * @return false Otherwise
+ *
+ * @note "Requesting csfle" means that it has set at least one search path OR
+ * has set the override path
+ */
+static bool
+_wants_csfle (mongocrypt_t *c)
+{
+   return c->opts.n_cselib_search_paths != 0 ||
+          c->opts.csfle_lib_override_path.data != NULL;
+}
+
+/**
+ * @brief Try to enable csfle for the given mongocrypt
+ *
+ * @param crypt The crypt object for which we should enable csfle
+ * @return true If no errors occurred
+ * @return false Otherwise
+ *
+ * @note Returns `true` even if loading fails to find the csfle library on the
+ * requested paths. `false` is only for hard-errors, which includes failure to
+ * load from the override path.
+ */
+static bool
+_try_enable_csfle (mongocrypt_t *crypt)
+{
+   mongocrypt_status_t *status = crypt->status;
+
+   _loaded_csfle found = _try_find_csfle (crypt);
 
    // If a CSFLE override path was specified, but we did not succeed in loading
    // CSFLE, that is a hard-error.
-   if (crypt->opts.csfle_lib_override_path.data &&
-       !mcr_dll_is_open (crypt->csfle_lib)) {
+   if (crypt->opts.csfle_lib_override_path.data && !found.okay) {
       CLIENT_ERR ("A CSFLE override path was specified [%s], but we failed to "
                   "open a dynamic library at that location",
                   crypt->opts.csfle_lib_override_path.data);
       return false;
    }
 
-   return true;
+   // Attempt to validate the try-find result against the global state:
+   const bool got_csfle =
+      _csfle_replace_or_take_validate_singleton (crypt, &found);
+   // Close the lib we found (may have been stolen in validate_singleton())
+   mcr_dll_close (found.lib);
+
+   if (got_csfle) {
+      crypt->csfle = g_csfle_state.vtable;
+      crypt->csfle_lib = g_csfle_state.csfle_lib;
+   }
+   // In cast of failure, validate_singleton() will set a non-ok status.
+   return mongocrypt_status_type (status) == MONGOCRYPT_STATUS_OK;
+}
+
+bool
+mongocrypt_init (mongocrypt_t *crypt)
+{
+   mongocrypt_status_t *status;
+
+   if (!crypt) {
+      return false;
+   }
+   status = crypt->status;
+   if (crypt->initialized) {
+      CLIENT_ERR ("already initialized");
+      return false;
+   }
+
+   crypt->initialized = true;
+
+   if (!mongocrypt_status_ok (crypt->status)) {
+      return false;
+   }
+
+   if (!_mongocrypt_opts_validate (&crypt->opts, status)) {
+      return false;
+   }
+
+   if (crypt->opts.log_fn) {
+      _mongocrypt_log_set_fn (
+         &crypt->log, crypt->opts.log_fn, crypt->opts.log_ctx);
+   }
+
+   if (!crypt->crypto) {
+#ifndef MONGOCRYPT_ENABLE_CRYPTO
+      CLIENT_ERR ("libmongocrypt built with native crypto disabled. crypto "
+                  "hooks required");
+      return false;
+#else
+      /* set default hooks. */
+      crypt->crypto = bson_malloc0 (sizeof (*crypt->crypto));
+      BSON_ASSERT (crypt->crypto);
+#endif
+   }
+
+   if (!_wants_csfle (crypt)) {
+      // User does not want csfle. Just succeed.
+      return true;
+   }
```

Comment:
This is vague in the scope document. The behavior I have in mind:

1. Try loading csfle from user provided paths
2. Otherwise if user did not provide paths, try loading csfle from system paths
3. Otherwise, use mongocryptd

Is it OK to modify `_try_find_csfle` search with `$SYSTEM` if neither `mongocrypt_setopt_append_csfle_search_path` or `mongocrypt_setopt_set_csfle_lib_path_override` were called?

Reply:
If that were the default behavior, how might a driver disable using the csfle library? My intuition is that "If you give me zero places to look, I will not look."

I feel like this option should be deferred to drivers, which can add `$SYSTEM` as the sole search path if they want that "default" behavior. e.g.:

```py
def init_mongocrypt(lib: MongoCryptLibrary, opts: InitOptions):
    if opts.enable_csfle:
        add_csfle_paths(lib)
    lib.init()

def add_csfle_paths(lib: MongoCryptLibrary, opts: InitOptions):
    for p in opts.paths_search:
        lib.add_csfle_search_path(p)
    p.add_csfle_search_path('$SYSTEM')
    for p in opts.paths_fallback:
        lib.add_csfle_search_path(p)
```

Reply:
Good point. Having drivers apply the default by adding `$SYSTEM` as the sole search path SGTM.